### PR TITLE
fix/custom-media-elements-timing: update mediaSetCallback to be entir…

### LINKED ISSUE #91: media-controller crashes when hot reloading codesandbox
1. Fork this codesandbox: https://codesandbox.io/s/mux-video-plus-media-chrome-html-sm8hf
2. Change command s to save
3. Page crashes when hot reloading

Seems to be a hot reloading issue because refreshing the embedded preview works.

hot-reload-bug_2021-11-10_11-46-00

### PR DESCRIPTION
**Pull request by cjpillsbury**
…ely promise based to avoid edge cases on custom elements being defined.

This PR refactors `mediaSetCallback` in `media-controller` and `media-container` to always use promises instead of internal promise + returning true/false. There were edge cases showing up in reloads on codesandbox (likely just one place exposing the edge cases) where a runtime error would occur because the custom element (in this case, `<mux-video/>`) wasn't yet ready, despite the promise+boolean implementation.

To test:
* start a local dev server and try loading the `<mux-video/>` example http://localhost:8080/examples/media-elements/mux-video.html
* Try throttling the network connection and clearing cache+reloading to validate potential timing issues.

NOTE: Since codesandbox doesn't currently support github URLs for npm dependencies, we cannot test in the original issue's env until after a version is published to npm with these changes. However, I have created a codesandbox with local versions of the relevant files here for testing: https://codesandbox.io/s/determined-zhukovsky-b3p06

* Fixes #91

### CODE REVIEW

File: src/js/media-container.js [hunk 1]
```diff
@@ -294,6 +307,29 @@ class MediaContainer extends window.HTMLElement {
...
+
+// Predicate that checks if a particular node minimally conforms to an HTMLMediaElement's shape.
+const isMediaEl = (node) => {
+  return (
+    node &&
+    mediaElPropTypeTuples.every(
+      ([propName, propType]) => typeof node[propName] === propType
+    )
+  );
+};
```
Comment (dylanjha) [inline src/js/media-container.js:331]: might this be better in `utils/`?

**Review by dylanjha — COMMENTED** [review]
Clearly I can see this covers:

1. change `mediaSetCallback` to return a promise instead of being sync
2. handles media elements that are custom elements (like `mux-video`) -- dealing with async nature of custom elements and when they get defined in the DOM
3. added an extra check for "does this look like a media element"

As far as I can tell (3) is new. Although it's not an exhaustive check, I think it's a good addition that can prevent folks trying to use a media element that is incompatible.

Overall looks good, might we be able to add some unit test coverage with this change?

File: src/js/media-container.js [hunk 2]
```diff
@@ -294,6 +307,29 @@ class MediaContainer extends window.HTMLElement {
   }
 }
 
+const mediaElPropTypes = {
```
[the remark this answers is not in the document]
Comment (dylanjha) [reply src/js/media-container.js:310]: That's probably more prudent. I can get behind checking for play/pause.

> Somewhere there's a line where we're sending a "this media does not support X feature" warning/error not a "this media is not compatible".

I also like this approach of not failing silently, but logging the missing APIs.

Can we do the following:

- Check for play/pause, have that be required
- Log the other missing APIs in a warning.

File: src/js/media-container.js [hunk 3]
```diff
@@ -196,33 +196,46 @@ class MediaContainer extends window.HTMLElement {
   }
 
   mediaSetCallback(media) {
-    // Should only ever be set with a compatible media element, never null
-    if (!media) {
-      console.error('<media-chrome>: Media element set with slot="media" does not appear to be compatible.', media);
-      return false;
-    }
+    const resolveMediaPromise = (media) => {
```
[the remark this answers is not in the document]
Comment (dylanjha) [reply src/js/media-container.js:199]: > mediaSetCallback was originally intended to mimic connectedCallback. As developer you're never responding to the result of connectedCallback, you're populating a known function property to be called when the element is connected.

Hmmm. Let me see if I understand your proposal:

- `media-container` handles the promise internally with a different function (like `handleMediaUpdated`)
- After media-container resolves the promise internally, then it calls `mediaSetCallback. If internal promise rejects then `mediaSetCallback` would never get called (instead of this proposed change which would reject the promise)
- so media-controller essentially doesn't need any refactoring because `mediaSetCallback` would no longer be returning a promise.

If I'm understanding this, it's all internal, I'm okay going either way with it.

[the remark this answers is not in the document]
Comment (heff) [reply src/js/media-container.js:199]: Yeah, exactly. We would just be removing the success check and `return` in media-controller's `if (!super.mediaSetCallback(media)) return;`, and anything that returns true or false in media-container's mediaSetCallback.

File: src/js/media-container.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (heff) [reply src/js/media-container.js:310]: Sounds good. Though unless it's simple and obvious, we should probably boot the second bullet to another issue or to-do, since this PR is part of solving a specific issue.

[the remark this answers is not in the document]
Comment (dylanjha) [reply src/js/media-container.js:310]: Just syncing with @cjpillsbury on this. I think this is a big more thorny.

We actually have a lot of code that relies on these properties for the internal state management (checking for `textTracks`, for example).

So essentially, this is not really valid use case. We've been pretty explicit from the beginning of "any media element must conform to the HTML Video element spec"

> Are there cases were you might want to be able to build a simple media element that just plays and pauses and just use the play button with it?

So now after talking it through, this "simple media element" isn't a valid use case unless we change our position that you don't need to use an element that conforms to the HTML Video element spec.

[the remark this answers is not in the document]
Comment (heff) [reply src/js/media-container.js]: Where's the line then? Does every media element have to implement the `.videoTracks` API and create their own version of media source extensions?

I don't know that just play/pause support is the right minimum, but requiring the entire video element API is unreasonable. We do want people to actually build media elements.

I think I'm more excited about making media container smart enough to handle when there's no textTracks property than making the bar to create a media element super high.

[the remark this answers is not in the document]
Comment (heff) [reply src/js/media-container.js]: I did try to call this out as a bigger topic. :) We don't protect against "unfinished" medias today, so this doesn't need to be a blocker to this PR, unless we want to make it one.

[the remark this answers is not in the document]
Comment (cjpillsbury) [reply src/js/media-container.js]: That's definitely fair. I guess I saw the async as "sidestepping" the issue and the interface check as an acceptable "solving" of the issue (at least disallowing the scenario to happen as a rule) based on prior conversations, but I'll go ahead and remove the structural type checking predication for now so it's not a blocker.

[the remark this answers is not in the document]
Comment (heff) [reply src/js/media-container.js]: That makes sense to me. Should we move it to a discussion?

**Review by heff — APPROVED** [review]
Awesome

File: src/js/media-container.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dylanjha) [reply src/js/media-container.js]: I think it can be an issue at this point, since we've discussed it and there's some action items

https://github.com/muxinc/media-chrome/issues/94
